### PR TITLE
Calibnet Support

### LINKED ISSUE #1320: Support for calibration network
Some issues in Forest can be tricky and time consuming to test when needing to fully sync a node. Adding support for the calibration network should allow us to easily sync nodes with similar parameters to mainnet to increase productivity in our testing workflow

AC
* Forest should sync to Filecoin's calibnet

Additional links
* Network parameters - https://docs.filecoin.io/networks/
* Lotus implementation - https://github.com/filecoin-project/lotus/blob/67bdc266afb47094ff7aca9adada7f768d2cf355/build/params_calibnet.go

The calibnet genesis and bootstrap peers can be found in Lotus' build directory under bootstrap for peer lists, and genesis for calibnet genesis file.

### PR DESCRIPTION
**Pull request by connormullett**
**Summary of changes**
Changes introduced in this pull request:
- Add support for calibration net

**Reference issue to close (if applicable)**

Closes #1320 

**Other information and links**

to build: `make calibnet`
calibnet snapshot can be made from lotus
Should fail after nv15 as changes are not introduced for the upgrade here

### CODE REVIEW

File: types/networks/src/calibnet/mod.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+pub const UPGRADE_ACTORS_V6_HEIGHT: ChainEpoch = 312746;
+
+pub const UPGRADE_PLACEHOLDER_HEIGHT: ChainEpoch = 9999999;
+
+/// Current network version for the network
+pub const NEWEST_NETWORK_VERSION: NetworkVersion = NetworkVersion::V14;
+
+/// Bootstrap peer ids
+pub const DEFAULT_BOOTSTRAP: &[&str] = &[
+    "/dns4/bootstrap-0.calibration.fildev.network/tcp/1347/p2p/12D3KooWRLZAseMo9h7fRD6ojn6YYDXHsBSavX5YmjBZ9ngtAEec",
```
Comment (ec2) [inline types/networks/src/calibnet/mod.rs:84]: Not sure where you go these, but they look different from https://github.com/filecoin-project/lotus/blob/master/build/bootstrap/calibnet.pi

File: types/networks/src/calibnet/mod.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    "/dns4/bootstrap-1.calibration.fildev.network/tcp/1347/p2p/12D3KooWJFtDXgZEQMEkjJPSrbfdvh2xfjVKrXeNFG1t8ioJXAzv",
+    "/dns4/bootstrap-2.calibration.fildev.network/tcp/1347/p2p/12D3KooWP1uB9Lo7yCA3S17TD4Y5wStP5Nk7Vqh53m8GsFjkyujD",
+    "/dns4/bootstrap-3.calibration.fildev.network/tcp/1347/p2p/12D3KooWLrPM4WPK1YRGPCUwndWcDX8GCYgms3DiuofUmxwvhMCn",
+];
+
+lazy_static! {
+    pub(super) static ref DRAND_SCHEDULE: [DrandPoint<'static>; 2] = [
```
Comment (ec2) [inline types/networks/src/calibnet/mod.rs:91]: Your Drand epoch schedule is wrong: https://github.com/filecoin-project/lotus/blob/master/build/params_calibnet.go#L15

File: vm/actor/src/builtin/sector.rs [hunk 3]
```diff
@@ -23,6 +23,9 @@ pub fn consensus_miner_min_power(p: RegisteredPoStProof) -> Result<StoragePower,
             if cfg!(feature = "interopnet") {
                 return Ok(StoragePower::from(2 << 30));
             }
+            if cfg!(feature = "calibnet") {
```
Comment (ec2) [inline vm/actor/src/builtin/sector.rs:26]: You're using this `calibnet` feature but you have not defined the feature within the crate's `Cargo.toml`.

File: types/networks/src/lib.rs [hunk 4]
```diff
@@ -15,41 +15,75 @@ mod drand;
...
 );
 
 #[cfg(all(
     feature = "mainnet",
-    any(feature = "interopnet", feature = "devnet", feature = "conformance",)
+    any(
+        feature = "interopnet",
+        feature = "devnet",
+        feature = "conformance",
+        feature = "calibnet"
```
Comment (ec2) [inline types/networks/src/lib.rs:30]: This feature is also not defined in the crate's `Cargo.toml`.

File: forest/Cargo.toml [hunk 5]
```diff
@@ -70,6 +70,7 @@ rocksdb = ["db/rocksdb", "ipld_blockstore/rocksdb"]
 sled = ["db/sled", "ipld_blockstore/sled"]
 devnet = ["actor/devnet", "networks/devnet", "testnet"]
 interopnet = ["actor/interopnet", "networks/interopnet", "testnet"]
+calibnet = ["networks/calibnet"]
```
Comment (ec2) [inline forest/Cargo.toml:73]: When you have defined the features in the `actor` and `networks` crate, you need to enable them here. 
```suggestion
calibnet = ["networks/calibnet", "networks/calibnet", "testnet"]
```

File: types/networks/src/calibnet/mod.rs (hunk 1 above)
Comment (connormullett) [reply types/networks/src/calibnet/mod.rs:84]: Looks like the ones I had are outdated, found them from filecoin docs

File: vm/actor/src/builtin/sector.rs (hunk 3 above)
Comment (connormullett) [reply vm/actor/src/builtin/sector.rs:26]: :+1:

File: types/networks/src/lib.rs (hunk 4 above)
Comment (connormullett) [reply types/networks/src/lib.rs:30]: :+1:

File: forest/Cargo.toml (hunk 5 above)
Comment (connormullett) [reply forest/Cargo.toml:73]: :+1:

File: types/networks/src/calibnet/mod.rs (hunk 2 above)
Comment (connormullett) [reply types/networks/src/calibnet/mod.rs:91]: Makes sense, last thing I would've checked

Should be

```
pub(super) static ref DRAND_SCHEDULE: [DrandPoint<'static>; 1] = [DrandPoint {
        height: 0,
        config: &*DRAND_MAINNET,
    },];
```
?

File: types/networks/Cargo.toml [hunk 6]
```diff
@@ -15,4 +15,5 @@ serde_json = "1.0"
 mainnet = []
 interopnet = []
 devnet = []
-conformance = []
\ No newline at end of file
+conformance = []
+calibnet = []
```
Comment (q9f) [inline types/networks/Cargo.toml:19]: You can configure your IDE to automatically insert new lines at the end of the file.

**Comment by elmattic** [comment]
> Is Calibnet ready to go? Could you rebase on main?

No it's not. There are still some sync issues.

**Comment by elmattic** [comment]
Hey @ec2  helping @connormullett with this sync issue and would like to know if you have some idea why this is failing.

Basically everything looks good during the block validation (height 636,822) after importing the snapshot except one call that internally failed.

The offending callstack goes like this:

https://github.com/ChainSafe/forest/blob/b202340f2acc1d70c4eef7491c8466895211386d/blockchain/chain_sync/src/tipset_syncer.rs#L1366

https://github.com/ChainSafe/forest/blob/b202340f2acc1d70c4eef7491c8466895211386d/blockchain/state_manager/src/lib.rs#L657

https://github.com/ChainSafe/forest/blob/b202340f2acc1d70c4eef7491c8466895211386d/blockchain/state_manager/src/lib.rs#L1263

https://github.com/ChainSafe/forest/blob/b202340f2acc1d70c4eef7491c8466895211386d/blockchain/state_manager/src/lib.rs#L138

https://github.com/ChainSafe/forest/blob/b202340f2acc1d70c4eef7491c8466895211386d/vm/state_tree/src/lib.rs#L216 (Fallbacking to v0 state tree since retrieval failed)

https://github.com/ChainSafe/forest/blob/b202340f2acc1d70c4eef7491c8466895211386d/vm/actor_interface/src/adt/map.rs#L39

...

https://github.com/ChainSafe/forest/blob/b202340f2acc1d70c4eef7491c8466895211386d/ipld/hamt/src/hamt.rs#L89

To sum up it seems to me that reading using the `Cid` `bafy2bzacec4xpti63qilepxw4ofgbt7pdduumnvufaom53sfrkz5xdcslfxdk` from the blockstore just returns a `None` and bubbles up to an error, later resulting in the log warning:

> 2022-02-15T11:38:30.664Z WARN  chain_sync::tipset_syncer            > Validating block [CID = bafy2bzacecnamj6hoxdsbdlkxs6fmug4nyndug2rubetymli6dkyf755gqeqg] in EPOCH = 636822 failed: Validation error: StateManager error: Cid (bafy2bzacec4xpti63qilepxw4ofgbt7pdduumnvufaom53sfrkz5xdcslfxdk) did not match any in database

Any ideas where this could come from or on how to continue debugging this? Thanks :)

**Comment by q9f** [comment]
linter fails on #1438

**Comment by connormullett** [comment]
Everything is g2g now. Should not sync past nv15 but doesn't panic

File: ipld/hamt/src/node.rs [hunk 7]
```diff
@@ -389,10 +389,11 @@ where
                 let deleted = child_node.rm_value(hashed_key, bit_width, depth + 1, key, store)?;
                 if deleted.is_some() {
                     *child = Pointer::Dirty(std::mem::take(child_node));
+
+                    // Clean to retrieve canonical form
+                    child.clean()?;
```
Comment (LesnyRumcajs) [inline ipld/hamt/src/node.rs:394]: Should this be backported to `ref-fvm`? If I understand correctly, they have the same issue on their side so calibnet can't work with any implementation using ref-fvm, right? https://github.com/filecoin-project/ref-fvm/blob/lemmih/forest/ipld/hamt/src/node.rs#L396

Comment (connormullett) [reply ipld/hamt/src/node.rs:394]: Good question, I'm going to say yes. I'll open a PR on their side with the same fix and see what they say

Comment (connormullett) [reply ipld/hamt/src/node.rs:394]: PR opened here: https://github.com/filecoin-project/ref-fvm/pull/343

Comment (LesnyRumcajs) [reply ipld/hamt/src/node.rs:394]: Thank you!

**Review by LesnyRumcajs — APPROVED** [review]

**Review by noot — APPROVED** [review]
worked for me! great work :D
